options/options.go

```diff
@@ -1838,105 +1881,138 @@ func (interp *interpreter) messageLiteralValue(
+				if ffld != nil {
+					// In editions, we support using the group type name only for fields that
+					// "look like" proto2 groups.
+					if protoreflect.Name(fieldNode.Name.Value()) == ffld.Message().Name() && // text format uses type name
+						ffld.Message().FullName().Parent() == ffld.FullName().Parent() && // message and field declared in same scope
+						ffld.Kind() == protoreflect.GroupKind /* uses delimited encoding */ {
+						// This one looks like a proto2 group, so it's a keeper.
+						err = nil
+					} else {
+						// It doesn't look like a proto2 group, so this is not a match.
+						ffld = nil
 					}
 				}
 			}
-			if errors.Is(err, protoregistry.NotFound) {
-				return protoreflect.Value{}, sourceinfo.OptionSourceInfo{},
-					reporter.Errorf(interp.nodeInfo(fieldNode.Name), "%vfield %s not found", mc, string(fieldNode.Name.Name.AsIdentifier()))
-			} else if err != nil {
-				return protoreflect.Value{}, sourceinfo.OptionSourceInfo{},
-					reporter.Error(interp.nodeInfo(fieldNode.Name), err)
+		}
+		if errors.Is(err, protoregistry.NotFound) {
+			err := interp.reporter.HandleErrorf(interp.nodeInfo(fieldNode.Name), "%vfield %s not found", mc, string(fieldNode.Name.Name.AsIdentifier()))
+			if err != nil {
+				return protoreflect.Value{}, sourceinfo.OptionSourceInfo{}, err
 			}
-			if fieldNode.Sep == nil && ffld.Message() == nil {
-				// If there is no separator, the field type should be a message.
-				// Otherwise, it is an error in the text format.
-				return protoreflect.Value{}, sourceinfo.OptionSourceInfo{},
-					reporter.Errorf(interp.nodeInfo(fieldNode.Val), "syntax error: unexpected value, expecting ':'")
+			hadError = true
+			continue
+		} else if err != nil {
+			err := interp.reporter.HandleErrorWithPos(interp.nodeInfo(fieldNode.Name), err)
+			if err != nil {
+				return protoreflect.Value{}, sourceinfo.OptionSourceInfo{}, err
 			}
-			srcInfo, err := interp.setOptionField(mc, msg, ffld, fieldNode.Name, fieldNode.Val, true, append(pathPrefix, int32(ffld.Number())))
+			hadError = true
+			continue
+		}
+		if err := interp.checkFieldUsage(targetType, ffld, fieldNode.Name); err != nil {
+			return protoreflect.Value{}, sourceinfo.OptionSourceInfo{}, err
+		}
+		if fieldNode.Sep == nil && ffld.Message() == nil {
+			// If there is no separator, the field type should be a message.
+			// Otherwise, it is an error in the text format.
+			err := interp.reporter.HandleErrorf(interp.nodeInfo(fieldNode.Val), "syntax error: unexpected value, expecting ':'")
 			if err != nil {
 				return protoreflect.Value{}, sourceinfo.OptionSourceInfo{}, err
 			}
+			hadError = true
+			continue
+		}
+		srcInfo, err := interp.setOptionField(targetType, mc, msg, ffld, fieldNode.Name, fieldNode.Val, true, append(pathPrefix, int32(ffld.Number())))
+		if err != nil {
+			return protoreflect.Value{}, sourceinfo.OptionSourceInfo{}, err
+		}
+		if srcInfo != nil {
 			flds[fieldNode] = srcInfo
 		}
 	}
+	if hadError {
```

Comment:
Could `interp.reporter` have a method to check if any errors were reported and avoid this state flag?

Reply:
There is such a method: [`*Handler.Error`](https://pkg.go.dev/github.com/bufbuild/protocompile@v0.10.0/reporter#Handler.Error). But an error could have been reported on a different option, prior to entering this function. There is a mode of this package that operates "leniently", where it can just interpret whatever it can, so we don't necessarily want to abort the results of this message value just because an error was reported for a different option.

There is also a [`*Handler.SubHandler`](https://pkg.go.dev/github.com/bufbuild/protocompile@v0.10.0/reporter#Handler.SubHandler) method, for creating a "child" of the handler, whose `Error` and `ReporterError` methods are limited to the errors reported through that child. But creating a sub-handler for every message literal processed feels too heavy-weight compared to adding a simple boolean flag.